fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -637,56 +619,68 @@ class ShardManager {
 				} else {
 					physicalShards[handle->GetName()] = std::make_shared<PhysicalShard>(db, handle->GetName(), handle);
 				}
 				columnFamilyMap[handle->GetID()] = handle;
-				TraceEvent(SevInfo, "ShardedRocksInitPhysicalShard", this->logId)
+				TraceEvent(SevVerbose, "ShardedRocksInitPhysicalShard", this->logId)
 				    .detail("PhysicalShard", handle->GetName());
 			}
+
 			KeyRange keyRange = prefixRange(shardMappingPrefix);
 			while (true) {
 				RangeResult metadata;
 				const int bytes = readRangeInDb(metadataShard.get(),
 				                                keyRange,
-				                                SERVER_KNOBS->ROCKSDB_READ_RANGE_ROW_LIMIT,
-				                                SERVER_KNOBS->ROCKSDB_READ_RANGE_ROW_LIMIT,
+				                                std::max(2, SERVER_KNOBS->ROCKSDB_READ_RANGE_ROW_LIMIT),
+				                                UINT16_MAX,
 				                                &metadata);
 				if (bytes <= 0) {
 					break;
 				}
 
-				std::vector<std::pair<KeyRange, std::string>> mapping =
-				    decodeShardMapping(metadata, shardMappingPrefix);
-
-				Key end = keyRange.begin.removePrefix(shardMappingPrefix);
-				for (const auto& [range, name] : mapping) {
-					TraceEvent(SevInfo, "ShardedRocksLoadRange", this->logId)
+				ASSERT_GT(metadata.size(), 0);
+				for (int i = 0; i < metadata.size() - 1; ++i) {
+					const std::string name = metadata[i].value.toString();
+					KeyRangeRef range(metadata[i].key.removePrefix(shardMappingPrefix),
+					                  metadata[i + 1].key.removePrefix(shardMappingPrefix));
+					TraceEvent(SevVerbose, "DecodeShardMapping", this->logId)
 					    .detail("Range", range)
-					    .detail("PhysicalShard", name)
-					    .log();
-					end = range.end;
+					    .detail("Name", name);
+
+					// Shard with an empty name is an empty shard.
 					if (name.empty()) {
 						continue;
 					}
+
 					auto it = physicalShards.find(name);
 					// Raise error if physical shard is missing.
 					if (it == physicalShards.end()) {
-						TraceEvent(SevError, "ShardedRocksDB").detail("MissingShard", name);
+						TraceEvent(SevError, "ShardedRocksDB", this->logId).detail("MissingShard", name);
 						return rocksdb::Status::NotFound();
 					}
+
 					std::unique_ptr<DataShard> dataShard = std::make_unique<DataShard>(range, it->second.get());
 					dataShardMap.insert(range, dataShard.get());
 					it->second->dataShards[range.begin.toString()] = std::move(dataShard);
 					activePhysicalShardIds.emplace(name);
 				}
-				if (end == specialKeys.end) {
+
+				if (metadata.back().key.removePrefix(shardMappingPrefix) == specialKeys.end) {
+					TraceEvent(SevVerbose, "ShardedRocksLoadShardMappingEnd", this->logId);
 					break;
 				}
-				keyRange = KeyRangeRef(end.withPrefix(shardMappingPrefix), keyRange.end);
+
+				if (metadata.size() == 1 && metadata.back().value.toString().empty()) {
+					// Should not happen, just being paranoid.
```

Comment:
This could only happen if someone accidentally changed the row limit to 1, right? Suggest putting some comments about the row limit either in the knob file or where the read happens. Also consider adding comments on why we need to read the last key again.

Reply:
The row limit is set to be >= 2 at line 639, so this should happen only when there is metadata corruption or other bugs.